### PR TITLE
Fix lookahead with None-delimited group

### LINKED ISSUE #84162: clang-sys 0.29.3 fails to build on nightly
I tried building `clang-sys 0.29.3` with feature `runtime` on the latest nightly.

I expected to see this happen: it builds successfully.

Instead, this happened: 
```console
> cargo +nightly build --features runtime |& head -n 20
   Compiling clang-sys v0.29.3 (/tmp/tmp.EH8ss9nTlb/clang-sys-0.29.3)
error[E0642]: patterns aren't allowed in methods without bodies
    --> src/lib.rs:1660:39
     |
1660 |     pub fn clang_CXCursorSet_contains(set: CXCursorSet, cursor: CXCursor) -> c_uint;
     |                                       ^^^
     |
help: give this argument a name or use an underscore to ignore it
     |
1660 |     pub fn clang_CXCursorSet_contains(_: CXCursorSet, cursor: CXCursor) -> c_uint;
     |                                       ^

error[E0642]: patterns aren't allowed in methods without bodies
    --> src/lib.rs:1660:57
     |
1660 |     pub fn clang_CXCursorSet_contains(set: CXCursorSet, cursor: CXCursor) -> c_uint;
     |                                                         ^^^^^^
     |
help: give this argument a name or use an underscore to ignore it
     |
```

### Meta[
`rustc --version --verbose`:
```
rustc 1.53.0-nightly (d0695c908 2021-04-12)
binary: rustc
commit-hash: d0695c9081b16077d0aed368bccaf437d77ff497
commit-date: 2021-04-12
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

### PR DESCRIPTION
**Pull request by Aaron1011**
Fixes https://github.com/rust-lang/rust/issues/84162, a regression introduced by https://github.com/rust-lang/rust/pull/82608.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/mod.rs [hunk 1]
```diff
@@ -929,16 +929,20 @@ impl<'a> Parser<'a> {
...
+        while i < dist {
+            token = cursor.next().0;
+            if matches!(
+                token.kind,
+                token::OpenDelim(token::NoDelim) | token::CloseDelim(token::NoDelim)
+            ) {
+                continue;
+            }
+            i += 1;
         }
```
Comment (pickfire) [inline compiler/rustc_parse/src/parser/mod.rs:944]:
```suggestion
        let mut token = Token::dummy();
        while dist > 0 {
            token = cursor.next().0;
            if !matches!(
                token.kind,
                token::OpenDelim(token::NoDelim) | token::CloseDelim(token::NoDelim)
            ) {
                dist -= 1;
            }
        }
```
Will this be better?

**Comment by Aaron1011** [comment]
Apparently, this is small performance *win* on a few benchmarks.

r? @petrochenkov

**Comment by Mark-Simulacrum** [comment]
rustc_parse compilation has regressed 10%, which is above the usual noise on that measurement. Is that expected?

**Comment by Aaron1011** [comment]
@Mark-Simulacrum I'm not sure - I don't understand anything about that benchmark run 🤷

**Comment by virtualritz** [comment]
Once this is merged, is it possible to kick off building `docs.rs` docs for all packages that rely on affected crates?
E.g. many `-sys` crates rely on `clang-sys` which is affected.

Or will crate authors have to publish bumped version crates just to get docs built in these cases?

**Comment by Aaron1011** [comment]
@virtualritz: It should be possible for affected crate authors to request rebuilds by opening issues on https://github.com/rust-lang/docs.rs

**Comment by jyn514** [comment]
@jethrogb this is tracked by https://github.com/rust-lang/rust/issues/84162, we don't prioritize PRs usually.

**Comment by jethrogb** [comment]
Whoops, my bad, it's right there in the PR description.

File: compiler/rustc_parse/src/parser/mod.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_parse/src/parser/mod.rs:963]: Looks less intuitive to me.

**Comment by rylev** [comment]
@Aaron1011 @petrochenkov @Mark-Simulacrum Looks like the [perf run](https://perf.rust-lang.org/compare.html?start=7537b20626100e7e7fc8c4ad3079d38c05338121&end=16bf626a31cb5b121d0bca2baa969b4f67eb0dab&stat=instructions:u) performed after this was merged showed the same regressions in compilation of the same benchmarks as well as compilation of `rustc_parse` crate. 

The compilation regressions in the benchmarks might be expected since this is strictly doing more work, but I wonder if we have an idea why this would regress compilation of the `rustc_parse` crate itself so much.

**Pull request merged**